src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.Windows.cs

```diff
@@ -7,10 +7,103 @@ namespace System.Net.Security
         }
+
+
+        private SslStreamCertificateContext(X509Certificate2 target, X509Certificate2[] intermediates)
+        {
+            if (intermediates.Length > 0)
+            {
+                using (X509Chain chain = new X509Chain())
+                {
+                    chain.ChainPolicy.VerificationFlags = X509VerificationFlags.AllFlags;
+                    chain.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;
+                    chain.ChainPolicy.DisableCertificateDownloads = true;
+                    bool osCanBuildChain = chain.Build(target);
+
+                    int count = 0;
+                    foreach (X509ChainStatus status in chain.ChainStatus)
+                    {
+                        if (status.Status.HasFlag(X509ChainStatusFlags.PartialChain) || status.Status.HasFlag(X509ChainStatusFlags.NotSignatureValid))
+                        {
+                            osCanBuildChain = false;
+                            break;
+                        }
+
+                        count++;
+                    }
+
+                    // OS failed to build the chain but we have at least some intermediates.
+                    // We will try to add them to "Intermediate Certification Authorities" store.
+                    if (!osCanBuildChain)
+                    {
+                        X509Store? store = new X509Store(StoreName.CertificateAuthority, StoreLocation.LocalMachine);
+
+                        try
+                        {
+                            store.Open(OpenFlags.ReadWrite);
+                        }
+                        catch
+                        {
+                            // If using system store fails, try to fall-back to user store.
+                            store.Dispose();
+                            store = new X509Store(StoreName.CertificateAuthority, StoreLocation.CurrentUser);
+                            try
+                            {
+                                store.Open(OpenFlags.ReadWrite);
+                            }
+                            catch
+                            {
+                                store.Dispose();
+                                store = null;
+                                if (NetEventSource.IsEnabled)
+                                {
+                                    NetEventSource.Error(this, $"Failed to open certificate store for intermediates.");
+                                }
+                            }
+                        }
+
+                        if (store != null)
+                        {
+                            using (store)
+                            {
+                                // Add everything except the root
+                                for (int index = count; index < intermediates.Length - 1; index++)
+                                {
+                                    store.Add(intermediates[index]);
```

Comment:
Is `intermediates` populated by the caller, or something already provided by doing a chain walk?  If it's caller data, then instead of adding things here (and assuming the order is semantic) you should add them to chain.ExtraStore, then after the chain build add things based on chain.ChainElements.

Reply:
This comes from the platform independent `chain.Build(target)`